File: backend/src/routes/calendar.ts
import express from 'express';
import jwt from 'jsonwebtoken';
import { calendarService } from '../services/calendar/calendarService.js';
import { authenticateToken } from '../middleware/auth.js';

const router = express.Router();

router.get('/oauth/url', authenticateToken, (req, res) => {
  try {
    const authHeader = req.headers.authorization || '';
    const jwtToken = authHeader.startsWith('Bearer ') ? authHeader.slice(7) : null;

    if (!jwtToken) {
      return res.status(400).json({ error: 'Missing bearer token' });
    }

    const url = calendarService.getOAuthUrl(jwtToken);
    return res.json({ url });
  } catch (err: any) {
    return res.status(500).json({ error: err.message });
  }
});

router.get('/oauth/callback', async (req, res) => {
  try {
    const code = req.query.code;
    if (!code) return res.status(400).json({ error: 'Missing code' });
    const state = req.query.state;
    if (!state) return res.status(400).json({ error: 'Missing state' });

    const JWT_SECRET = process.env.JWT_SECRET;
    if (!JWT_SECRET) return res.status(500).json({ error: 'JWT_SECRET not configured' });

    let userId: string;
    try {
      const payload: any = jwt.verify(state as string, JWT_SECRET);
      userId = payload.sub || payload.userId;
    } catch {
      return res.status(401).json({ error: 'Invalid state token' });
    }

    await calendarService.exchangeCodeAndStoreTokens(code as string, userId);

    const frontendUrl = process.env.FRONTEND_URL || 'http://localhost:5173';
    return res.redirect(302, `${frontendUrl}/trips?gcal=connected`);
  } catch (err: any) {
    const frontendUrl = process.env.FRONTEND_URL || 'http://localhost:5173';
    return res.redirect(302, `${frontendUrl}/trips?gcal=error&reason=${encodeURIComponent(err.message || 'unknown')}`);
  }
});

router.get('/calendar/upcoming', authenticateToken, async (req, res) => {
  try {
    const events = await calendarService.listUpcomingEvents(req.userId);
    return res.json({ events });
  } catch (err: any) {
    if (err.message === 'Google not connected') {
      return res.status(401).json({ error: 'Google not connected' });
    }
    return res.status(500).json({ error: err.message });
  }
});

router.post('/calendar/events', authenticateToken, async (req, res) => {
  try {
    const { summary, description, location, start, end, timeZone } = req.body;

    if (!summary || !start || !end) {
      return res.status(400).json({ error: 'summary, start, and end are required' });
    }

    const event = await calendarService.createEvent(req.userId, {
      summary,
      description,
      location,
      start,
      end,
      timeZone,
    });

    return res.status(201).json({ event });
  } catch (err: any) {
    if (err.message === 'Google not connected') {
      return res.status(401).json({ error: 'Google not connected' });
    }
    return res.status(500).json({ error: err.message });
  }
});

export default router;

File: backend-python/app/routes/calendar.py
"""Calendar routes — Google OAuth + event management."""

import jwt
from fastapi import APIRouter, Depends, HTTPException, Query, Request
from fastapi.responses import RedirectResponse

from app.config import settings
from app.deps import get_current_user
from app.models import User
from app.services.calendar_service import calendar_service

router = APIRouter()


@router.get("/oauth/url")
async def get_oauth_url(request: Request, user: User = Depends(get_current_user)):
    auth_header = request.headers.get("Authorization", "")
    token = auth_header.replace("Bearer ", "") if auth_header.startswith("Bearer ") else None
    if not token:
        raise HTTPException(status_code=400, detail="Missing bearer token")
    url = calendar_service.get_oauth_url(token)
    return {"url": url}


@router.get("/oauth/callback")
async def oauth_callback(
    code: str = Query(...),
    state: str = Query(...),
):
    try:
        payload = jwt.decode(state, settings.jwt_secret, algorithms=["HS256"])
        user_id = payload.get("sub") or payload.get("userId")
        if not user_id:
            raise HTTPException(status_code=401, detail="Invalid state token")
    except jwt.InvalidTokenError:
        raise HTTPException(status_code=401, detail="Invalid state token")

    await calendar_service.exchange_code_and_store_tokens(code, str(user_id))
    return RedirectResponse(
        url=f"{settings.frontend_url}/trips?gcal=connected",
        status_code=302,
    )


@router.get("/calendar/upcoming")
async def upcoming_events(user: User = Depends(get_current_user)):
    try:
        events = await calendar_service.list_upcoming_events(str(user.id))
        return {"events": events}
    except ValueError as e:
        raise HTTPException(status_code=401, detail=str(e))


@router.post("/calendar/events")
async def create_event(
    payload: dict,
    user: User = Depends(get_current_user),
):
    required = ["summary", "start", "end"]
    for field in required:
        if field not in payload:
            raise HTTPException(status_code=400, detail=f"{field} is required")
    try:
        event = await calendar_service.create_event(str(user.id), payload)
        return {"event": event}
    except ValueError as e:
        raise HTTPException(status_code=401, detail=str(e))
